sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/DatasetManager.scala

```diff
@@ -0,0 +1,289 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pipelines.graph
+
+import scala.jdk.CollectionConverters._
+import scala.util.control.{NonFatal, NoStackTrace}
+
+import org.apache.spark.SparkException
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.connector.catalog.{
+  CatalogV2Util,
+  Identifier,
+  TableCatalog,
+  TableChange,
+  TableInfo
+}
+import org.apache.spark.sql.connector.expressions.Expressions
+import org.apache.spark.sql.pipelines.graph.QueryOrigin.ExceptionHelpers
+import org.apache.spark.sql.pipelines.util.SchemaInferenceUtils.diffSchemas
+import org.apache.spark.sql.pipelines.util.SchemaMergingUtils
+
+/**
+ * [[DatasetManager]] is responsible for materializing tables in the catalog based on the given
+ * graph. For each table in the graph, it will create a table if none exists (or if this is a
+ * full refresh), or merge the schema of an existing table to match the new flows writing to it.
+ */
+object DatasetManager extends Logging {
+
+  /**
+   * Wraps table materialization exceptions.
+   *
+   * The target use case of this exception is merely as a means to capture attribution -
+   * 1. Indicate that the exception is associated with table materialization.
+   * 2. Indicate which table materialization failed for.
+   *
+   * @param tableName The name of the table that failed to materialize.
+   * @param cause The underlying exception that caused the materialization to fail.
+   */
+  case class TableMaterializationException(
+      tableName: String,
+      cause: Throwable
+  ) extends Exception(cause)
+      with NoStackTrace
+
+  /**
+   * Materializes the tables in the given graph. This method will create or update the tables
+   * in the catalog based on the given graph and context.
+   *
+   * @param resolvedDataflowGraph The resolved [[DataflowGraph]] with resolved [[Flow]] sorted
+   *                              in topological order.
+   * @param context The context for the pipeline update.
+   * @return The graph with materialized tables.
+   */
+  def materializeDatasets(
+      resolvedDataflowGraph: DataflowGraph,
+      context: PipelineUpdateContext
+  ): DataflowGraph = {
+    val (_, refreshTableIdentsSet, fullRefreshTableIdentsSet) = {
+      DatasetManager.constructFullRefreshSet(resolvedDataflowGraph.tables, context)
+    }
```

Comment:
refresh tables are tables needs to be updated during the current execution, by default it should be all but users can configure if they only update a subset of the table.

full fresh tables are tables needs to be completely recomputed, we will wipe out all the historical data and start from the new. This by default is None but can also be specified by users which tables they wanna full refresh